### PR TITLE
[5.7] Honor PSR-16 SimpleCache interface

### PR DESCRIPTION
**Pull request by plehatron**
Cache repository methods `set` and `setMultiple`, defined in the SimpleCache interface do not return any value, but they should according to PSR-16. This PR aims to solve this and fix issue laravel/framework#26674.

Cache stores Apc, Array, Database, File, Memcached, Null, and Redis are affected. Their respective unit tests were updated to check for return values.

### CODE REVIEW

File: src/Illuminate/Cache/DatabaseStore.php [hunk 1]
```diff
@@ -100,9 +100,9 @@ public function put($key, $value, $minutes)
         $expiration = $this->getTime() + (int) ($minutes * 60);
 
         try {
-            $this->table()->insert(compact('key', 'value', 'expiration'));
+            return $this->table()->insert(compact('key', 'value', 'expiration'));
         } catch (Exception $e) {
-            $this->table()->where('key', $key)->update(compact('value', 'expiration'));
+            return $this->table()->where('key', $key)->update(compact('value', 'expiration'));
```
Comment (X-Coder264) [inline src/Illuminate/Cache/DatabaseStore.php]: This returns an integer representing the number of rows affected by the query so this has to be casted to bool.

**Comment by X-Coder264** [comment]
You've changed the docblocks from `void` to `bool` on the concrete `Repository` implementation but you left them to be void on the interface. Also there was no need to change the `forever` method return type to honor the PSR-16 interface as that method is not on that interface to begin with. Generally changing a method signature is considered a breaking change (especially now that it happened on a contract) so this PR should probably target master (as all Store implementations out there will not honor the interface anymore after this gets merged).

btw, pozdrav iz Trikodera 😄

File: src/Illuminate/Cache/Repository.php [hunk 2]
```diff
@@ -194,55 +194,63 @@ public function pull($key, $default = null)
...
 
-            return;
+            return $result;
         }
 
         if (! is_null($minutes = $this->getMinutes($minutes))) {
-            $this->store->put($this->itemKey($key), $value, $minutes);
+            $result = $this->store->put($this->itemKey($key), $value, $minutes);
 
             $this->event(new KeyWritten($key, $value, $minutes));
```
Comment (X-Coder264) [inline src/Illuminate/Cache/Repository.php]: Technically we now have the result variable to tell us if it was written successfully so this event shouldn't be fired if `$result === false`.

File: src/Illuminate/Cache/ArrayStore.php [hunk 3]
```diff
@@ -32,11 +32,13 @@ public function get($key)
...
      * @param  mixed   $value
      * @param  float|int  $minutes
-     * @return void
+     * @return bool
      */
     public function put($key, $value, $minutes)
     {
         $this->storage[$key] = $value;
+
+        return array_key_exists($key, $this->storage);
```
Comment (X-Coder264) [inline src/Illuminate/Cache/ArrayStore.php]: In what case can this be false?

**Comment by taylorotwell** [comment]
What are we accomplishing for users for using PSR-16 at all?

**Comment by X-Coder264** [comment]
@taylorotwell  I'm using some framework-agnostic packages which depend on PSR-16 on a couple of Laravel projects and having the ability to inject the Illuminate Cache Repository there is a great thing. Since we already claim to implement that interface I don't see the harm in actually fixing this so that we are compliant with it, especially since it's not a huge change.

**Comment by plehatron** [comment]
@taylorotwell In my case, I'm working on a project where we use the [jeremykendall/php-domain-parser](https://github.com/jeremykendall/php-domain-parser) package. It has a dependency for any PSR-16 compatible cache library. As reported in issue #26674, caching with Laravel's cache component doesn't work because the package strictly expects the return values for certain methods to be boolean (as defined in the PSR-16 contract). What I had to do is create a wrapper around the cache component and basically fix the return values to make it work, which kind of defeats the purpose of having a cache component that tries to be PSR-16 complaint.

With this PR we're aiming to make the cache component fully compliant with PSR-16, so there are no surprises, and workarounds have to be made for packages that expect strict return types.

I can't say exactly what percentage of users will benefit from this, but I can say that the adoption of PSR-16 is growing. When `SimpleCache` interface was implemented with PR #20194, there were 120 packages depending on psr/simple-cache. Today there are [413 packages](https://packagist.org/packages/psr/simple-cache/dependents), that's a 244% increase in almost 1.5 years.

**Comment by plehatron** [comment]
> You've changed the docblocks from `void` to `bool` on the concrete `Repository` implementation but you left them to be void on the interface. Also there was no need to change the `forever` method return type to honor the PSR-16 interface as that method is not on that interface to begin with. Generally changing a method signature is considered a breaking change (especially now that it happened on a contract) so this PR should probably target master (as all Store implementations out there will not honor the interface anymore after this gets merged).
> 
> btw, pozdrav iz Trikodera 😄

@X-Coder264 Good catch, thanks. Fixed it with last changes. I figured since `forever` internally uses `put` it can just use it's return value since it's now available. Also, wanted to make it a bit cleaner, now when you look at the `Illuminate\Contracts\Cache\Store`, none of the methods return `void`.

ps, pozdrav s ex3k strane 👋 😃

**Comment by taylorotwell** [comment]
Thanks for the info.

**Comment by taylorotwell** [comment]
Do we actually implement the PSR-16 interfaces anywhere?

**Comment by deleugpn** [comment]
I think the only PR regarding psr-16 was https://github.com/laravel/framework/pull/20194.

**Pull request closed without merging**